### DISCUSSION TITLE
Questions about the two methods of insert_textbox or insert_text

### DISCUSSION THREAD

**Discussion by zidong-onepiece1** [pymupdf/PyMuPDF#1633] [Q&A]
When I use `get_text('json')` to parse out the content of the pdf, I get the **flags** of the text at the span level.
But when I use the` insert_text` or `insert_textbox `method to insert the text into the new pdf, the **flags** parameter or similar parameters are not found, so how can I reproduce the original attributes of the text (such as italics, bold, etc.)

**Comment by JorjMcKie** [comment]
Except for the flag indicating a superscript, the other flags express properties of the font in use for the span, like bold, or mono.
So if you use that font again, you will automatically also use those flags.

Note that "using that font again" may not be as simple as it sounds, because you may need to find its xref first, then extract it and insert it again in the target page.
There are a few interesting discussions here dealing with this topic.

**Reply by zidong-onepiece1** [reply]
Thanks for your prompt reply! My original intention of raising this question is to do full-text translation of PDF, for example, to directly translate English PDF files into Chinese PDF files, so the text processing needs to change the font, but retain the features of the font.

I studied the `insert_text` and` insert_textbox` methods again and found that there is a parameter called **oc**, maybe I can start from the parameter **oc.**

**Reply by JorjMcKie** [reply]
> I studied the insert_text and insert_textbox methods again and found that there is a parameter called oc, maybe I can start from the parameter oc.

No ``oc`` has nothing to do with it.
I am sure know better about Chinese fonts in general. But you probably need some type of a list of fonts with the needed property combinations.
Maybe you should simplify this aspect a little at the beginning and just use the general fallback font in PyMuPDF "Droid Sans Fallback Regular", which supports Chinese among other languages. Once you have solved all the issues around correct text positioning and text color, you can take a second round and now also find the right Chinese font for a Latin font that for example is "mono, bold and italic".

One of the problems you will encounter is what should be done if the Chinese translation is **_longer_** than the original. Decrease the fontsize? You will probably never get translated text that has the same length as the original.

I would also use classes `TextWriter` and `Font` instead of `insert_text()` / ``insert_textbox()``. Those two classes have a number of advantages, as explained in the documentation.
The general approach may be along this snippet

```python
# ipage = input page, Latin original
# opage = output page in Chinese, created with same width / height
# myflags set to text extraction without images:
myflags = fitz.TEXTFLAGS_DICT & ~fitz.TEXT_PRESERVE_IMAGES
# textwriter for Chinese output text:
tw = fitz.TextWriter(opage.rect)
# font for Chinese text (example: Droid Sans Fallback Regular):
font = fitz.Font("cjk")
for block in ipage.get_text("dict", flags=myflags)["blocks"]:
    for line in block["lines"]:
        for span in line["spans"]:
            fontsize = span["size"]  # fontsize
            latin = span["text"]  # English original text
            ilength = fitz.Rect(span["bbox"]).width  # length of English text on page
            chinese = translate(latin)  # invoke your translater function
            # compute length of Chinese text using output font
            olength = font.text_length(chinese, fontsize=fontsize)
            # now decide what to do in cases ilength < olength etc.
            # to simply write the translated text at the same position:
            pos = fitz.Point(span["origin"])  # insertion point input page
            tw.append(pos, chinese, font=font, fontsize=fontsize)

# now all Chinese text is stored in the textwriter, put it out on output page:
tw.write_text(opage)

# When done for all input pages:
outpdf.subset_fonts()  # build font subsets to remove unused characters from fonts
outpdf.save("translated.pdf", garbage=4, deflate=True, ...)
```

**Reply by zidong-onepiece1** [reply]
Yes, I also noticed the classes `TextWriter `and `Font `you mentioned. Thanks again for your guidance！

> One of the problems you will encounter is what should be done if the Chinese translation is longer than the original. Decrease the fontsize? You will probably never get translated text that has the same length as the original.

Regarding the problem, reducing the font size is the easiest way, but will result in a lot of white space in the new PDF. The method I use is to automatically adapt to the size of the rect according to the text, select the font size to fill as much as possible into the rect, but this method will cause the font size of different blocks in the same page to be inconsistent, and the font size needs to be unified.

**Reply by JorjMcKie** [reply]
> but this method will cause the font size of different blocks in the same page to be inconsistent, ...

Yes I can image the effect. Not trivial. How to deal with this and find a satisfactory solution is probably more important than finding fonts with specialized properties.
